internal/vault/auth_methods.go

```diff
@@ -0,0 +1,475 @@
+	Description      string                 `json:"description"`
+	UsersOrGroupsKey string                 `json:"usersOrGroupsKey"`
+	Roles            []interface{}          `json:"roles"`
+	Users            interface{}            `json:"users"`
+	Crossaccountrole []interface{}          `json:"crossaccountrole"`
+	Groups           map[string]interface{} `json:"groups"`
+	Options          map[string]interface{} `json:"options"`
+	Map              map[string]interface{} `json:"map"`
+	Config           map[string]interface{} `json:"config"`
+}
+
+// getExistingAuthMethods gets all auth methods that are already in Vault.
+// The existing auth methods are in a map to make it easy to disable easily from it with "O(n)" complexity.
+func (v *vault) getExistingAuthMethods() (map[string]*api.MountOutput, error) {
+	existingAuths := make(map[string]*api.MountOutput)
+
+	existingAuthList, err := v.cl.Sys().ListAuth()
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to list existing auth methods")
+	}
+
+	for _, existingAuthMethod := range existingAuthList {
+		existingAuths[existingAuthMethod.Type] = existingAuthMethod
+	}
+
+	return existingAuths, nil
+}
+
+// getUnmanagedAuthMethods gets unmanaged auth methods by comparing what's already in Vault and what's in the externalConfig
+func (v *vault) getUnmanagedAuthMethods(managedAuthMethods []auth) map[string]*api.MountOutput {
+	// TODO: remove this line and just use passed value
+	unmangedAuths, _ := v.getExistingAuthMethods()
+	// Remove managed auth methods form the items since the rest will be disabled.
+	for _, mangedAuthMethod := range managedAuthMethods {
+		delete(unmangedAuths, mangedAuthMethod.Type)
+	}
+	// Remove token auth method since it's the default
+	delete(unmangedAuths, "token")
+
+	return unmangedAuths
+}
+
+func (v *vault) configureAuthMethods() error {
+	managedAuths := extConfig.Auth
+	existingAuths, _ := v.getExistingAuthMethods()
+	unManagedAuths := v.getUnmanagedAuthMethods(managedAuths)
+
+	err := v.syncManagedAuthMethods(managedAuths, existingAuths)
+	if err != nil {
+		return errors.Wrap(err, "error configuring managed auth methods")
+	}
+
+	err = v.disableUnmanagedAuthMethods(unManagedAuths)
+	if err != nil {
+		return errors.Wrap(err, "error while disabling unmanaged auth methods")
+	}
+
+	return nil
+}
+
+// Disables any auth method that's not managed if purgeUnmanagedConfig option is enabled
+func (v *vault) disableUnmanagedAuthMethods(unManagedAuths map[string]*api.MountOutput) error {
+	if len(unManagedAuths) == 0 || !extConfig.PurgeUnmanagedConfig.Enabled || extConfig.PurgeUnmanagedConfig.Exclude.Auths {
+		return nil
+	}
+
+	for authMethod := range unManagedAuths {
+		err := v.cl.Sys().DisableAuth(authMethod)
+		if err != nil {
+			return errors.Wrapf(err, "error disabling %s auth method in vault", authMethod)
+		}
+	}
+	return nil
+}
+
+func (v *vault) syncManagedAuthMethods(managedAuths []auth, existingAuths map[string]*api.MountOutput) error {
```

Comment:
I'd suggest sticking to `add` and `remove` instead of `sync` and `disable` to keep it consistent with the other config (e.g. policies, secrets engines, etc.)

also using `remove` makes it explicit disabling an auth is actually removing.